source/_posts/2016-09-30-spark-streaming-on-yarn.markdown

```diff
@@ -0,0 +1,324 @@
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue
+
+
+Another important issue for Spark Streaming job is keeping processing time stable and highly predictable.
+Processing time should stay below batch duration to avoid delays. 
+I've found that Spark speculative execution helps a lot, especially on busy cluster. 
+Batch processing times are much more stable than when speculative execution is disabled.
+Unfortunately speculative mode can be enabled only if Spark actions are idempotent.
+
+    spark-submit --master yarn --deploy-mode cluster \
+        --conf spark.yarn.maxAppAttempts=4 \
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue \
+        --conf spark.speculation=true
+
+
+## Security
+
+On a secured HDFS cluster, long-running Spark Streaming jobs fails due to Kerberos ticket expiration.
+Without additional settings, Kerberos ticket is issued when Spark Streaming job is submitted to the cluster.
+When ticket expires Spark Streaming job is not able to write or read data from HDFS anymore.
+ 
+In theory (based on documentation) it should be enough to pass Kerberos principal and keytab as spark-submit command:
+ 
+    spark-submit --master yarn --deploy-mode cluster \
+         --conf spark.yarn.maxAppAttempts=4 \
+         --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+         --conf spark.yarn.max.executor.failures={8 * num_executors} \
+         --conf spark.yarn.executor.failuresValidityInterval=1h \
+         --conf spark.task.maxFailures=8 \
+         --queue realtime_queue \
+         --conf spark.speculation=true \
+         --principal user/hostname@domain \
+         --keytab /path/to/foo.keytab
+
+In practice, due to several bugs ([HDFS-9276](https://issues.apache.org/jira/browse/HDFS-9276), [SPARK-11182](https://issues.apache.org/jira/browse/SPARK-11182))
+HDFS cache must be disabled. If not, Spark will not be able to read updated token from file on HDFS.
+
+    spark-submit --master yarn --deploy-mode cluster \
+         --conf spark.yarn.maxAppAttempts=4 \
+         --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+         --conf spark.yarn.max.executor.failures={8 * num_executors} \
+         --conf spark.yarn.executor.failuresValidityInterval=1h \
+         --conf spark.task.maxFailures=8 \
+         --queue realtime_queue \
+         --conf spark.speculation=true \
+         --principal user/hostname@domain \
+         --keytab /path/to/foo.keytab \
+         --conf spark.hadoop.fs.hdfs.impl.disable.cache=true
+
+
+## Logging
+
+The easiest way for Spark jobs to access application logs is to configure Log4j console appender, 
```

Comment:
"The easiest way to access Spark job logs..." - teraz brzmi jak by to Spark szukał swoich logów